src/shogun/features/DotFeatures.cpp

```diff
@@ -93,9 +90,9 @@ void CDotFeatures::dense_dot_range(float64_t* output, int32_t start, int32_t sto
 #endif
 		{
 			if (alphas)
-				output[i]=alphas[i]*this->dense_dot(i, vec, dim)+b;
+				output[i]=alphas[i]*this->dense_dot(i + start, vec, dim)+b;
```

Comment:
I think `start` is included earlier.

To make sure everything is fine, have you tested if with different number of threads?

Reply:
`i` ranges from `t_start` to `t_stop`. But `t_start` begins from `0` and `t_stop` goes up to `stop`.
It should be that `t_start` begins from `0` and `t_stop` goes up to `stop - start`, or `t_start` begins from `start` and `t_stop` goes up to `stop`.

Reply:
Judging from this comment
``` C++
// write access is internally between output[start..stop] so the following
// line is necessary to write to output[0...(stop-start-1)]
output-=start;
```
This was initially written so that we go from `start` to  `end`, so we subtract `start` from the output, so it transforms to `0` to `end - start`.

Reply:
This assumption didn't hold, as `i` ranges from `0` to `t_stop`
```C++
int32_t t_start=thread_num*step;
int32_t t_stop=(thread_num==num_threads) ? stop : (thread_num+1)*step;
```
And even if it did, the indexing of `alpha` would be wrong.

Reply:
Should I revert the changes and write a simple test that fails?

Reply:
That sounds like a good idea. You can add it to another branch if that's more efficient than reverting. Link to test results or copy+paste suffices if that is faster as well.

`t_start` starts on a different value depending on the thread.

Reply:
I ran this test on the old version, which tests the underflow:
```C++
TEST_F(DotFeaturesTest, dense_dot_range)
{
	index_t num_feats = 2;
	index_t num_vectors = 4;

	SGMatrix<float64_t> data(num_feats, num_vectors);
	for (index_t i = 0; i < num_vectors; i++)
		data.get_column(i).set_const(5);
	auto feats = some<CDenseFeatures<float64_t>>(data);

	SGVector<float64_t> vec(num_feats);
	vec.set_const(3);

	index_t start = 1;
	index_t stop = num_vectors;
	
	float64_t* buffer = new float64_t[num_vectors];
	float64_t& gaurd = buffer[0];
	gaurd = 0;
	float64_t* output = buffer  + 1;

	feats->dense_dot_range(
	    output, start, stop, nullptr, vec.vector, num_feats, 0);

	ASSERT_EQ(gaurd, 0);
}
```
Here `gaurd` should not be changed at all.
These are the results:
```
$ ./shogun-unit-test --gtest_filter=DotFeaturesTest.dense_dot_range
Note: Google Test filter = DotFeaturesTest.dense_dot_range
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DotFeaturesTest
[ RUN      ] DotFeaturesTest.dense_dot_range
/home/theartful/Programming/shogun/tests/unit/features/DotFeatures_unittest.cc:149: Failure
      Expected: gaurd
      Which is: 30
To be equal to: 0
[  FAILED  ] DotFeaturesTest.dense_dot_range (12 ms)
[----------] 1 test from DotFeaturesTest (12 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (15 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DotFeaturesTest.dense_dot_range

 1 FAILED TEST
```

Reply:
Memory leak with `buffer`?
In any case, use std::vector instead.

Reply:
The test leaks indeed. Freeing the memory doesn't change the test result. The problem is that the old `dense_dot_range` changes `output[-1]`

Reply:
All right, the point is clear now. Thanks for writing this test, good illustration.